api/agent/protocol/json.go

```diff
@@ -0,0 +1,133 @@
+package protocol
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+// This is sent into the function
+// All HTTP request headers should be set in env
+type JSONIO struct {
+	Headers    http.Header `json:"headers,omitempty"`
+	Body       string      `json:"body"`
+	StatusCode int         `json:"status_code,omitempty"`
+}
+
+// JSONProtocol converts stdin/stdout streams from HTTP into JSON format.
+type JSONProtocol struct {
+	in  io.Writer
+	out io.Reader
+}
+
+func (p *JSONProtocol) IsStreamable() bool {
+	return true
+}
+
+type RequestEncoder struct {
+	*json.Encoder
+}
+
+func (h *JSONProtocol) DumpJSON(w io.Writer, req *http.Request) error {
+	stdin := json.NewEncoder(h.in)
+	_, err := io.WriteString(h.in, `{`)
+	if err != nil {
+		// this shouldn't happen
+		return err
+	}
+	if req.Body != nil {
```

Comment:
i don't think body will ever be nil outside of any test cases written, but it seems like including it as an empty string is probably preferable to omitting it as part of our defined api here

Reply:
i.e. an `http.Request` will always have a non-nil body, even if it's of length 0

Reply:
So, you're saying that there's no need check body at this place, just attempt to read it and see if it's an empty string?

Reply:
seems like it does what you would expect: https://play.golang.org/p/FTBdbti-OA -- so yes

Reply:
no need to check, just encode buffer